File: src/org/genyris/interp/builtin/ReplaceCarFunction.java
package org.genyris.interp.builtin;

import org.genyris.core.Exp;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.GenyrisException;

public class ReplaceCarFunction extends ApplicableFunction {

	public ReplaceCarFunction(Interpreter interp) {
		super(interp);
	}

	public Exp bindAndExecute(Closure proc, Exp[] argument, Environment envForBindOperations) throws GenyrisException {
		if( argument.length != 2)
			throw new GenyrisException("Too many or few arguments to rplaca: " + argument.length);
		return argument[0].setCar(argument[1]);
	}

}

File: src/org/genyris/interp/builtin/ReplaceCarFunction.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp.builtin

import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.interp.ApplicableFunction
import org.genyris.interp.Closure
import org.genyris.interp.Environment
import org.genyris.interp.Interpreter

class ReplaceCarFunction(interp: Interpreter) : ApplicableFunction(interp, "rplaca", true) {
    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(proc: Closure?, argument: Array<Exp?>, envForBindOperations: Environment?): Exp? {
        checkArguments(argument, 2)
        return argument[0]!!.setCar(argument[1])
    }
}
